### ISSUE TITLE
Clarify whether 'any'/'LSPAny' fields are optional

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode-languageserver-node#970]
In v3.16 of the spec, some items were defined as `any`:

```js
interface DidChangeConfigurationParams {
	/**
	 * The actual changed settings
	 */
	settings: any;
}
```

My interpretation was that `undefined` was value here (which means they can be omitted).

In v3.17, they are now `LSPAny`:

```js
interface DidChangeConfigurationParams {
	/**
	 * The actual changed settings
	 */
	settings: LSPAny;
}
```

Where `LSPAny` is defined as:

```ts
/**
 * The LSP any type
 *
 * @since 3.17.0
 */
export type LSPAny = LSPObject | LSPArray | string | integer | uinteger | decimal | boolean | null;
```

And in the meta model the field is not marked as `optional`:

```js
{
	"name": "settings",
	"type": {
		"kind": "reference",
		"name": "LSPAny"
	},
	"documentation": "The actual changed settings"
}
```

It's not clear to me whether this field is intended to be optional or not. In my original TS parsing I treated `any` as being optional, but in my new meta model code that only happens if `optional: true` so there's a difference in my generated code.

@dbaeumer should `LSPAny` be treated as always optional? If so, is it possible the meta model could explicitly mark them as `optional` to avoid consumers needing to special case (or even know about) this?

**Comment by dbaeumer** [comment]
This misinterpretation was actually the reason why we introduced `LSPAny`. The specification uses JSON typed as TypeScript and JSON doesn't support an undefined value for a property. Only omitted properties. This is why the spec uses `?` instead of `| undefined` since it means something different. Even in TS now.

To answer your question: no LSPAny shouldn't be treated as always optional. If the property is optional we need to mark it as optional.

**Comment by dbaeumer** [comment]
LSPAny does include `null` which the VS Code client sends in case of no settings.

**Comment by DanTup** [comment]
@dbaeumer so to clarify, both of the following are _required_ (not optional)? `DidChangeConfigurationParams.settings` is marked with a `?` in the published 3.16 version of the spec but is not marked as optional in 3.17/meta model.

- `DidChangeConfigurationParams.settings`
- `ProgressParams.value`

In case you missed it, the only other outstanding issue I have is about the definition of `T`:

https://github.com/microsoft/vscode-languageserver-node/issues/946#issuecomment-1134561500

Should this be dropped from the meta_model and replaced with `LSPAny` or something?
